### DISCUSSION TITLE
aarch64 support

### DISCUSSION THREAD

**Discussion by stumpylog** [zxing-cpp/zxing-cpp#536] [Ideas]
I'd like to use this in a project which supports armv7, aarch64 and x86_64.  The armv7 wheel should be buildable with PiWheels now, so that leaves aarch64 as a remaining arch I'd like to not need to build for myself.

`cibuildwheel` supports building wheels for aarch64 through emulation.  But emulation is slower, so the would make the wheel building slower.  With some other jobs in the workflow, it probably wouldn't add that much time, though that does add some complication and duplication unfortunately.

So before I explore too much, is there any desire for adding wheels for this arch?

**Comment by axxel** [comment]
Just today I added cibuildwheel based support for macos arm64 builds. When you talk about aarch64 you mean linux/android arm64 builds?

The runtime of the build-python action is already way too long for me to let it run regularly, therefore it is only started manually / or on a release. I would not mind if it took say 30%-50% longer. On the other hand, a simple `pip install zxing-cpp` should work with a pre-build wheel or without one, right?

**Reply by stumpylog** [reply]
Right, I was thinking of linux/aarch64.

I'll look into how much time it adds.  I'm guessing it'll be a good chunk thanks to the emulation.  Installing from source with `pip` isn't really a viable solution when building our Docker image, mostly thanks to the likely time to compile under emulation.  But maybe every so often isn't too bad in the release flow here.

**Reply by stumpylog** [reply]
I experimented.  It unfortunately adds almost an hour to the dist building, which is a massive jump.  I'll figure something out instead.  Thanks for the feedback!

**Reply by axxel** [reply]
That is 'unpleasant', indeed. Spending this time on every docker-image setup on your end is even worse, I get that. I'm not totally against adding this and waiting that time during my release preparation. But I'm also wondering why there is no better alternative. When building the module for the arm64 architecture on macos this is achieved via cross compilation. Why is that not an option on linux?

**Reply by axxel** [reply]
FYI: I just spent an hour running down that rabbit hole: https://github.com/pypa/cibuildwheel/issues/598. It looked very promising at first but seems to have simply been abandoned for no obvious reason. :-(.

**Reply by stumpylog** [reply]
It's obviously more work, but another option is going native.  Oracle's cloud includes native ARM instances in the free tier: https://www.oracle.com/cloud/compute/arm/.  Using a self hosted runner there to build just aarch64, upload the artifact wheels and finally upload the results to pypi could be a solution.  There have been reports though of Oracle not being so nice a neighbor and disabling/removing instances with little to no warning. 

Or a Raspberry Pi could also be a self hosted runner for ARM64. It's been the default install I believe for a while now.

**Reply by fr567** [reply]
I would also be interested in a solution here.

**Comment by stumpylog** [comment]
It looking like you could use Cirrus combined with GHA to build natively, while not needing to manage a self hosted instance.  I stumbled across the updates `pikepdf` made for building it's aarch wheels.

Something like this https://github.com/pikepdf/pikepdf/blob/main/.cirrus.yml combined with a wait like this https://github.com/pikepdf/pikepdf/blob/main/.github/workflows/build.yml#L49 looks to end up with wheels, built natively, then transferred to GHA for actual uploading to pypi.

If that seems ok, I can look into it more

**Reply by axxel** [reply]
Looking at the runtimes of the cirrus ci setup for pikepdf (https://github.com/pikepdf/pikepdf/actions/workflows/build.yml) of randomly taking 30 to 120 minutes does not look very promising with respect to the goal of having manageable built times. I question the cost benefit ratio of adding another CI tool in that case. Seeing specific results might be convincing, so, if you want to give it a try, please do so.

As I was mentioning [here](https://github.com/zxing-cpp/zxing-cpp/issues/552#issuecomment-1516382558) I stumbled upon some caching technique related to cibuildwheel that should allow to significantly reduce the total build time by compiling the c++ files only once and use it for all 3 currently supported python versions. Unfortunately, I have not been able to find proper information about that. Maybe someone else is more successful.

**Comment by oeiber** [comment]
I'm using paperless-ngx, which is using zxing-cpp for proper barcode detection. The software runs on a raspberry pi 5 (aarch64). But building aarch64 wheels is causing a lot of time and cpu usage.
So it would be great if there would be native support for aarch64.

**Comment by stumpylog** [comment]
GitHub now has ARM64 runner options for both Windows and Ubuntu images.  I haven't looked at the ci for this in quite some time, but it seems that should be a much easier method to produce native wheels

**Reply by axxel** [reply]
Thank you very much for the hint. I'm low on time a.t.m., so if you should have some PR ready ideas on how to incorporate that, feel invited to let me know ;).

**Reply by axxel** [reply]
Update: current HEAD builds native arm wheels for macOS and Linux (for a while now). Can be downloaded e.g. [here](https://github.com/zxing-cpp/zxing-cpp/actions/runs/20799307497/attempts/1#summary-59740561331). Will be published to pypi.org with the upcoming 3.0 release.

**Reply by stumpylog** [reply]
Awesome to hear.  Sorry I didn't get time to investigate it more myself

**Comment by axxel** [comment]
There are now native arm packages for linux and mac on pypi.org: https://pypi.org/project/zxing-cpp/#files. Hence, I am closing this discussion.
